### PR TITLE
HBW-270 Get tasks by entities lazy

### PR DESCRIPTION
**Pull request by NatalieKor**
Feb-07-2020 18-33-52

### CODE REVIEW

File: hbw/app/javascript/packs/hbw/components/container.js.jsx [hunk 1]
```diff
@@ -24,6 +24,46 @@ modulejs.define(
         return tasks.filter(task => task.entity_code === entityCode).reverse();
       };
 
+      getTasksFormsByList = async (tasks) => {
```
Comment (GideonEXE) [inline hbw/app/javascript/packs/hbw/components/container.js.jsx]: This methods should be in the store.

File: hbw/app/javascript/packs/hbw/components/container.js.jsx [hunk 2]
```diff
@@ -24,6 +24,46 @@ modulejs.define(
...
+        const emptyFormTasks = this.getEmptyFormEntityTasks(tasks);
+        const { connection } = this.props.env;
+        const { addForms } = this.context;
+
+        if (emptyFormTasks.length > 0) {
+          const forms = await connection.request({
```
Comment (GideonEXE) [inline hbw/app/javascript/packs/hbw/components/container.js.jsx]: If you will use destruction here, then you can avoid  forms.forms in  mergeFormsToTasks

File: hbw/app/javascript/packs/hbw/components/container.js.jsx [hunk 3]
```diff
@@ -24,6 +24,46 @@ modulejs.define(
...
+          storeTask => entityTasks.filter(
+            entityTask => entityTask.id === storeTask.id
+          ).length > 0
+        );
+
+        const emptyFormEntityTasks = entityTasksFromStore.filter(
+          entityTaskFromStore => (entityTaskFromStore.form.fields === undefined)
+        );
+
+        Object.keys(emptyFormEntityTasks).forEach((field) => {
```
Comment (GideonEXE) [inline hbw/app/javascript/packs/hbw/components/container.js.jsx]: Object.entries

File: hbw/app/javascript/packs/hbw/components/shared/context/store.jsx [hunk 4]
```diff
@@ -43,6 +43,21 @@ const withStoreContext = (WrappedComponent) => {
       return incomingTask;
     };
 
+    mergeFormsToTasks = (prevTasks, forms) => {
+      const updatedTasks = prevTasks;
+      Object.keys(forms.forms).forEach((field) => {
```
Comment (GideonEXE) [inline hbw/app/javascript/packs/hbw/components/shared/context/store.jsx]: Object.entries

**Comment by NatalieKor** [comment]
In tests I found three different tasks with equal id by one entity, so plan to choose task BY responce form was failed (just in case of this wrong data. Every task must have different id). Now we have to choose form by task.

File: hbw/app/javascript/packs/hbw/components/shared/context/store.jsx [hunk 5]
```diff
@@ -66,6 +66,58 @@ const withStoreContext = (WrappedComponent) => {
...
+    getEmptyFormEntityTasks = (entityTasks) => {
+      const entityClass = this.props.env.entity_class;
+      const tasksIds = [];
+
+      const emptyFormEntityTasks = entityTasks.filter(
+        entityTaskFromStore => (entityTaskFromStore.form.fields === undefined)
+      );
+
+      // eslint-disable-next-line no-unused-vars
+      Object.entries(emptyFormEntityTasks).forEach(([_, task]) => {
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/shared/context/store.jsx]: `map`

File: hbw/app/javascript/packs/hbw/components/entity_task.js.jsx [hunk 6]
```diff
@@ -57,21 +61,24 @@ modulejs.define(
 
       renderForm = () => {
         const { form } = this.props.task;
-
-        if (this.state.error) {
-          return <Error error={this.state.error} env={this.props.env} />;
+        if (form.variables !== undefined) {
```
[the remark this answers is not in the document]
Comment (NatalieKor) [reply hbw/app/javascript/packs/hbw/components/entity_task.js.jsx]: `form` when it's "empty" has key `csrf_token`, so check here for nothing on form is impossible. The only way to check really empty form is check `form.variables`. Or I didn't understand your point.

File: hbw/app/javascript/packs/hbw/components/shared/context/store.jsx [hunk 7]
```diff
@@ -67,6 +67,60 @@ const withStoreContext = (WrappedComponent) => {
...
+    mergeFormsToTasks = (prevTasks, fetchedForms) => {
+      const tasks = prevTasks;
+      const forms = fetchedForms;
+
+      Object.values(tasks).forEach((value) => {
+        const fetchedForm = forms.find(form => form.task_id === value.id);
+
+        if (fetchedForm !== undefined) {
+          value.form = fetchedForm.form_fields;
+          value.form.csrf_token = fetchedForm.csrf_token;
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/shared/context/store.jsx]: Let's try to return form from backend with proper format.

File: hbw/app/javascript/packs/hbw/components/shared/context/store.jsx [hunk 8]
```diff
@@ -67,6 +67,60 @@ const withStoreContext = (WrappedComponent) => {
...
+      const entityClass = this.props.env.entity_class;
+      const tasksIds = [];
+
+      const emptyFormEntityTasks = entityTasks.filter(
+        task => (task.form === undefined)
+      );
+
+      Object.values(emptyFormEntityTasks).forEach((task) => {
+        tasksIds.push({ task_id: task.id, entity_class: entityClass });
+      });
```
Comment (TimAle) [inline hbw/app/javascript/packs/hbw/components/shared/context/store.jsx]: forEach + push = map

File: hbw/app/models/hbw/common/adapter.rb [hunk 9]
```diff
@@ -89,6 +85,18 @@ def form(task_id, entity_class)
         end
       end
 
+      def get_forms_for_task_list(tasks)
+        tasks = JSON.parse(tasks)
```
Comment (TimAle) [inline hbw/app/models/hbw/common/adapter.rb]: Do not overwrite input arguments. You can just do `forms = JSON.parse(tasks).map do |task|` here.

File: hbw/app/views/hbw/tasks/_task.json.jbuilder [hunk 10]
```diff
@@ -10,4 +10,4 @@ json.set! :process_name, task.process_name
 json.set! :process_instance_id, task.process_instance_id
 json.set! :entity_code, task.entity_code(params[:entity_class].to_sym)
 json.set! :entity_url, entity_url(task, params[:entity_class].to_sym)
-json.set! :form, task.form.as_json.merge(csrf_token: csrf_token)
+json.set! :form, task.form.as_json.merge(csrf_token: csrf_token) if !task.form.nil?
```
Comment (TimAle) [inline hbw/app/views/hbw/tasks/_task.json.jbuilder]: `if !` -> `unless`

**Pull request merged**